nomos-mix/core/src/persistent_transmission.rs

```diff
@@ -26,6 +30,86 @@ impl Default for PersistentTransmissionSettings {
     }
 }
 
+pub struct PersistentTransmissionStream<S>
+where
+    S: Stream,
+{
+    interval: Interval,
+    coin: Coin<ChaCha12Rng>,
+    queue: VecDeque<S::Item>,
+    stream: S,
+}
+
+impl<S> PersistentTransmissionStream<S>
+where
+    S: Stream,
+{
+    pub fn new(
+        settings: PersistentTransmissionSettings,
+        stream: S,
+    ) -> PersistentTransmissionStream<S> {
+        let interval = time::interval(Duration::from_secs_f64(
+            1.0 / settings.max_emission_frequency,
+        ));
+        let coin = Coin::<_>::new(
+            ChaCha12Rng::from_entropy(),
+            settings.drop_message_probability,
+        )
+        .unwrap();
+        Self {
+            interval,
+            coin,
+            queue: Default::default(),
+            stream,
+        }
+    }
+}
+
+impl<S> Stream for PersistentTransmissionStream<S>
+where
+    S: Stream<Item = Vec<u8>> + Unpin,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let Self {
+            ref mut interval,
+            ref mut queue,
+            ref mut stream,
+            ref mut coin,
+            ..
+        } = self.get_mut();
+        if pin!(interval).poll_tick(cx).is_ready() {
+            match queue.pop_front() {
+                item @ Some(_) => return Poll::Ready(item),
+                None => {
+                    if coin.flip() {
+                        return Poll::Ready(Some(DROP_MESSAGE.to_vec()));
+                    }
+                }
+            };
+        };
+        if let Poll::Ready(Some(item)) = pin!(stream).poll_next(cx) {
+            queue.push_back(item);
+        }
+        Poll::Pending
+    }
+}
```

Comment:
What do you think about removing the `queue: VecDeque` and use the `stream` as a queue, as below? The test passes, but I'd like to ask if I'm missing something.
```rust
impl<S> Stream for PersistentTransmissionStream<S>
where
    S: Stream<Item = Vec<u8>> + Unpin,
{
    type Item = S::Item;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let Self {
            ref mut interval,
            ref mut stream,
            ref mut coin,
            ..
        } = self.get_mut();
        if pin!(interval).poll_tick(cx).is_ready() {
            if let Poll::Ready(Some(item)) = pin!(stream).poll_next(cx) {
                return Poll::Ready(Some(item));
            } else if coin.flip() {
                return Poll::Ready(Some(DROP_MESSAGE.to_vec()));
            }
        };
        Poll::Pending
    }
}
```

Reply:
So, this depends on the "backpresure" law that we want to have. For example, we could have a stream wrapper over a bounded channel in where if we do not pull and we block the buffer it will block sends. 
I actually just tried to maintain the original implementation here. But I think that we could go for removing the queue.

Reply:
Ah, you're right. My suggestion works only when the user uses an unbound channel. The user doesn't care about the internal implementation of the `PersistentTransmissionStream`, which delays consuming messages from the channel. So, yeah, it makes sense to consume messages from the channel (stream) and store them in the queue.

Reply:
I just saw your commit [12f134c](https://github.com/logos-co/nomos-node/pull/860/commits/12f134c7369d9f65fdb6d0d636da49fc3b7f1351) that removed the queue. I think we can go with the way that you think the optimal. I want to learn the practice.

Reply:
Yeha, I think that's the issue that most of backpressures have. The producers don't know how fast the consumer will consume the channel.